### PR TITLE
Field Reference enhancements

### PR DESCRIPTION
**Pull request by jacques-n**
- Nest FieldReference into the expression message (since it is an expression)
- Add a definition of root_type for FieldReference to allow use field selections against expressions, not just root structs.
- Remove unusually complex selection subtypes (map name pattern, list range)

### CODE REVIEW

File: binary/expression.proto [hunk 1]
```diff
@@ -216,6 +215,101 @@ message Expression {
       repeated string prerequisite = 2;
     }
   }
+
+  message ReferenceSegment {
```
Comment (cpcloud) [inline binary/expression.proto:219]: Can you add a small piece of documentation to the toplevel messages here? I think that's `ReferenceSegment`, `MaskExpression` and `FieldReference` if I'm reading it correctly.

File: binary/expression.proto [hunk 2]
```diff
@@ -216,6 +215,132 @@ message Expression {
       repeated string prerequisite = 2;
     }
   }
+
+  // A way to reference the inner property of a complex record. Can reference 
+  // either a map key by literal, a struct field by the ordinal position of 
+  // the desired field or a particular element in an array. Supports
+  // expressions that would roughly translate to something similar to:
+  // a.b[2].c['my_map_key'].x where a,b,c and x are struct field references 
+  // (ordinalized in the internal representation here), [2] is an list offset
```
Comment (cpcloud) [inline binary/expression.proto:224]: small nit: `an list` -> `a list`

File: binary/expression.proto [hunk 3]
```diff
@@ -216,6 +215,132 @@ message Expression {
...
+      // Optional child segment
+      ReferenceSegment child = 2;
+    }
+
+  }
+
+  // A reference that takes an existing subtype and selectively removes fields from 
+  // it. For example, one might initially have an inner struct with 100 fields but a 
+  // a particular operation only needs to interact with only 2 of those 100 fields.
+  // In this situation, one would you a mask expression to eliminate the 98 fields that
```
Comment (cpcloud) [inline binary/expression.proto:263]: `you` -> `use`?

File: binary/expression.proto [hunk 4]
```diff
@@ -251,3 +376,5 @@ message AggregateFunction {
   AggregationPhase phase = 4;
   Type output_type = 5;
 }
+
+
```
Comment (cpcloud) [inline binary/expression.proto:380]: Kill these newlines

**Review by cpcloud — DISMISSED** [review]
LGTM, small typo nits.

**Review by cpcloud — APPROVED** [review]

File: binary/expression.proto (hunk 4 above)
Comment (jacques-n) [reply binary/expression.proto]: I'm going to keep adding these just to motivate someone to add more checks :D

**Pull request merged**